### PR TITLE
feat(controller): Retry transient offload errors. Resolves #4464

### LINKED ISSUE #4464: Transient database errors with offload enabled cause workflow to fail
## Summary 

If there is a transient database connection error with offload enabled, then active workflows are marked as failed. I would at least expect some kind of retry.

## Diagnostics

What Kubernetes provider are you using? 
GKE

What version of Argo Workflows are you running? 
2.11.0-rc1

```
Paste the logs from the workflow controller:
kubectl logs -n argo $(kubectl get pods -l app=workflow-controller -n argo -o name) | grep ${workflow}
```

```
 time="2020-11-02T13:32:42Z" level=error msg="hydration failed: dial tcp 10.X.X.X:5432: connect: connection refused" namespace=default workflow=XXXX
 time="2020-11-02T13:32:42Z" level=error msg="Failed to archive workflow" err="dial tcp 10.X.X.X:5432: connect: connection refused" namespace=default workflow=XXXX
```

Seems to be caused by controller.go:517 which marks any workflows as 'error' if hydration fails.

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by alexec**
Resolves #4464

Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

**Comment by alexec** [comment]
No need for this, already done in : https://github.com/argoproj/argo/commit/ee848921348676718a8ab4cef8e8c2f52b86d124

**Comment by alexec** [comment]
`TestWFDefaultWithWFTAndWf`

**Comment by alexec** [comment]
A user has tested this change successfully.

File: workflow/controller/operator.go [hunk 1]
```diff
@@ -1899,7 +1899,11 @@ func (woc *wfOperationCtx) markWorkflowFailed(message string) {
 }
 
 func (woc *wfOperationCtx) markWorkflowError(err error) {
-	woc.markWorkflowPhase(wfv1.NodeError, err.Error())
+	if errorsutil.IsTransientErr(err) {
+		woc.markWorkflowPhase(woc.wf.Status.Phase, fmt.Sprintf("workflow had transient error: %v", err))
```
Comment (alexec) [inline workflow/controller/operator.go:1903]: this is a significant change - I think this makes this v3 feature

**Comment by alexec** [comment]
I think this is the wrong approach now.

**Comment by alexec** [comment]
@sarabala1979 I've re-opened this PR with fewer changes (i.e. easier to review). I would probably backport this to v2.11.9

**Review by sarabala1979 — APPROVED** [review]

**Comment by alexec** [comment]
Merged into https://github.com/argoproj/argo/releases/tag/v2.12.0-rc4

**Pull request merged**